corehq/util/couch_doc_processor.py

```diff
@@ -367,3 +382,47 @@ def _processing_complete(self):
         ))
         if self.skipped:
             print(DOCS_SKIPPED_WARNING.format(self.skipped))
+
+
+class BulkDocProcessor(CouchDocumentProcessor):
+    """Process couch docs in batches
+
+    The bulk doc processor will send a batch of documents to the document
+    processor. If the processor does not respond with True then
+    the iteration is halted. Restarting the iteration will start by
+    re-sending the previous chunk to the processor
```

Comment:
I think this statement about re-sending the previous chunk on restart is not always true. Scenario:

- retrieve a doc, we'll call it *the first doc*, `should_process` returns true, doc added to `changes` list
- retrieve number of docs > chunk size; `should_process` happens to return false for all of them. By now the state of the underlying `ResumableDocsByTypeIterator` has advanced and *the first doc* will not be yielded again *unless* it is `retry()`d.
- retrieve enough docs to make `changes` list length == chunk size, `should_process` returns true for all of them
- `_process_chunk` calls `process_bulk_docs`, which returns false and halts the iteration.
- next time the process is run, the first set of changes sent to the processor is not the same as the previous chunk that had been sent.